File: packages/core/src/params.ts
import { PolyLLMError } from './errors.js'
import type { ModelSpec, ParamPolicy, UnifiedRequest } from './types.js'

export interface ParamValidationResult {
  request: UnifiedRequest
  warnings: string[]
}

function unsupported(
  policy: ParamPolicy,
  name: string,
  reason: string,
  warnings: string[],
): void {
  if (policy === 'strict') {
    throw new PolyLLMError('UNSUPPORTED_PARAMETER', `参数不被目标模型支持: ${name}。${reason}`, {
      parameter: name,
      reason,
      policy,
    })
  }
  warnings.push(`${name} 已忽略: ${reason}`)
}

function clampNumber(
  name: string,
  value: number,
  min: number,
  max: number,
  policy: ParamPolicy,
  warnings: string[],
): number {
  if (value < min || value > max) {
    const bounded = Math.min(max, Math.max(min, value))
    if (policy === 'auto') {
      warnings.push(`${name}=${value} 超出范围 [${min}, ${max}]，已调整为 ${bounded}`)
      return bounded
    }
    if (policy === 'lenient') {
      warnings.push(`${name}=${value} 超出范围 [${min}, ${max}]，已忽略`)
      return min
    }
    throw new PolyLLMError('UNSUPPORTED_PARAMETER', `${name}=${value} 超出范围 [${min}, ${max}]`, {
      parameter: name,
      value,
      min,
      max,
    })
  }
  return value
}

export function validateAndMapParams(
  model: ModelSpec,
  request: UnifiedRequest,
  policy: ParamPolicy = 'strict',
): ParamValidationResult {
  return validateAndMap(model, request, policy)
}

function validateAndMap(
  model: ModelSpec,
  request: UnifiedRequest,
  policy: ParamPolicy,
): ParamValidationResult {
  const warnings: string[] = []
  const next: UnifiedRequest = {
    ...request,
    model: model.id,
    messages: request.messages,
  }

  if (request.stream && !model.capabilities.streaming) {
    unsupported(policy, 'stream', '该模型不支持流式输出', warnings)
    next.stream = false
  }
  if (request.jsonMode && !model.capabilities.jsonMode) {
    unsupported(policy, 'jsonMode', '该模型不支持 JSON mode', warnings)
    next.jsonMode = false
  }
  if (request.tools?.length && !model.capabilities.functionCalling) {
    unsupported(policy, 'tools', '该模型不支持 function calling', warnings)
    next.tools = []
  }
  if (!model.capabilities.systemPrompt && request.messages.some((message) => message.role === 'system')) {
    unsupported(policy, 'messages[role=system]', '该模型不支持 system prompt', warnings)
    next.messages = request.messages.filter((message) => message.role !== 'system')
  }
  if (!model.capabilities.vision) {
    const hasImage = request.messages.some(
      (message) => Array.isArray(message.content) && message.content.some((part) => part.type === 'image_url'),
    )
    if (hasImage) unsupported(policy, 'messages[content=image_url]', '该模型不支持视觉输入', warnings)
    next.messages = next.messages.map((message) =>
      Array.isArray(message.content)
        ? { ...message, content: message.content.filter((part) => part.type !== 'image_url') }
        : message,
    )
  }

  if (request.temperature !== undefined && model.params.temperature) {
    next.temperature = clampNumber(
      'temperature',
      request.temperature,
      model.params.temperature.min,
      model.params.temperature.max,
      policy,
      warnings,
    )
  }
  if (request.topP !== undefined && model.params.topP) {
    next.topP = clampNumber('topP', request.topP, model.params.topP.min, model.params.topP.max, policy, warnings)
  }
  if (request.maxTokens !== undefined) {
    const max = model.params.maxTokens?.max ?? model.maxOutput
    next.maxTokens = clampNumber('maxTokens', request.maxTokens, 1, max, policy, warnings)
  }

  return { request: next, warnings }
}

File: python/packages/core/src/polyllm_core/params.py
from __future__ import annotations

from dataclasses import replace
from typing import Any, Mapping

from .errors import PolyLLMError
from .types import ModelSpec, ParamPolicy, UnifiedMessage, UnifiedRequest


class ParamValidationResult:
    def __init__(self, request: UnifiedRequest, warnings: list[str]):
        self.request = request
        self.warnings = warnings


def validate_and_map_params(
    model: ModelSpec,
    request: UnifiedRequest,
    policy: ParamPolicy = "strict",
) -> ParamValidationResult:
    warnings: list[str] = []

    def unsupported(name: str, reason: str) -> bool:
        if policy == "strict":
            raise PolyLLMError(
                "UNSUPPORTED_PARAMETER",
                f"参数不被目标模型支持: {name}。{reason}",
                {"parameter": name, "reason": reason, "policy": policy},
            )
        warnings.append(f"{name} 已忽略: {reason}")
        return False

    def clamp_number(name: str, value: float, minimum: float, maximum: float) -> float:
        if minimum <= value <= maximum:
            return value
        bounded = min(maximum, max(minimum, value))
        if policy == "auto":
            warnings.append(f"{name}={value} 超出范围 [{minimum}, {maximum}]，已调整为 {bounded}")
            return bounded
        if policy == "lenient":
            warnings.append(f"{name}={value} 超出范围 [{minimum}, {maximum}]，已忽略")
            return minimum
        raise PolyLLMError(
            "UNSUPPORTED_PARAMETER",
            f"{name}={value} 超出范围 [{minimum}, {maximum}]",
            {"parameter": name, "value": value, "min": minimum, "max": maximum},
        )

    stream = request.stream
    if stream and not model.capabilities.streaming:
        stream = unsupported("stream", "该模型不支持流式输出")

    json_mode = request.json_mode
    if json_mode and not model.capabilities.json_mode:
        json_mode = unsupported("jsonMode", "该模型不支持 JSON mode")

    tools = list(request.tools)
    if tools and not model.capabilities.function_calling:
        tools = [] if unsupported("tools", "该模型不支持 function calling") else tools

    messages = list(request.messages)
    if not model.capabilities.system_prompt and any(message.role == "system" for message in messages):
        keep_system = unsupported("messages[role=system]", "该模型不支持 system prompt")
        messages = [message for message in messages if keep_system or message.role != "system"]

    if not model.capabilities.vision:
        has_image = any(
            not isinstance(message.content, str) and any(part.type == "image_url" for part in message.content)
            for message in messages
        )
        if has_image:
            keep_images = unsupported("messages[content=image_url]", "该模型不支持视觉输入")
            messages = [
                message
                if isinstance(message.content, str) or keep_images
                else replace(message, content=[part for part in message.content if part.type != "image_url"])
                for message in messages
            ]

    temperature = request.temperature
    temperature_range = model.params.get("temperature")
    if temperature is not None and temperature_range:
        temperature = clamp_number("temperature", temperature, temperature_range.min, temperature_range.max)

    top_p = request.top_p
    top_p_range = model.params.get("top_p")
    if top_p is not None and top_p_range:
        top_p = clamp_number("topP", top_p, top_p_range.min, top_p_range.max)

    max_tokens = request.max_tokens
    if max_tokens is not None:
        max_tokens_range = model.params.get("max_tokens")
        maximum = max_tokens_range.max if max_tokens_range else model.max_output
        max_tokens = int(clamp_number("maxTokens", max_tokens, 1, maximum))

    prepared = replace(
        request,
        model=model.id,
        messages=messages,
        stream=stream,
        json_mode=json_mode,
        tools=tools,
        temperature=temperature,
        top_p=top_p,
        max_tokens=max_tokens,
    )
    return ParamValidationResult(prepared, warnings)
